src/solvers/adaptive_time_solver.C

```diff
@@ -86,15 +87,38 @@ void AdaptiveTimeSolver::reinit()
 
 void AdaptiveTimeSolver::advance_timestep ()
 {
+  // The first access of advance_timestep happens via solve, not user code
+  // It is used here to store any initial conditions data
+  if (!first_solve)
+    {
+      // We call advance_timestep in user code after solve, so any solutions
+      // we will be storing will be for the next time instance
+      _system.time += _system.deltat;
+    }
+    else
+    {
+      // We are here because of a call to advance_timestep that happens
+      // via solve, the very first solve. All we are doing here is storing
+      // the initial condition. The actual solution computed via this solve
+      // will be stored when we call advance_timestep in the user's timestep loop
+      first_solve = false;
+    }
+
+  // If the user has attached a memory or file solution history object
+  // to the solver, this will store the current solution indexed with
+  // the current time
+  solution_history->store(false);
+
   NumericVector<Number> & old_nonlinear_soln =
     _system.get_vector("_old_nonlinear_solution");
   NumericVector<Number> & nonlinear_solution =
     *(_system.solution);
 
   old_nonlinear_soln = nonlinear_solution;
 
-  if (!first_solve)
-    _system.time += last_deltat;
```

Comment:
Wait - how was this ever working before, if we didn't have a ```first_solve = false``` case until just now?

Reply:
Was it that we were always setting first_solve = false in TwoStepTimeSolver?

Reply:
Yup.